tests/smoke_tests/run_smoke_test.py

```diff
@@ -0,0 +1,182 @@
+) -> None:
+    with open(config_path, "r") as file:
+        config = yaml.safe_load(file)
+
+    _preload_dataset(dataset_path, config)
+
+    # Start the server and capture its process object
+    logger.info("Starting server...")
+    server_process = await asyncio.create_subprocess_exec(
+        "nohup",
+        "python",
+        "-m",
+        "examples.fedprox_example.server",
+        "--config_path",
+        config_path,
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.STDOUT,
+    )
+
+    # reads lines from the server output in search of the startup log message
+    # times out after 20s of inactivity if it doesn't find the log message
+    full_server_output = ""
+    startup_message = "FL starting"
+    output_found = False
+    while True:
+        try:
+            assert server_process.stdout is not None, "Server's process stdout is None"
+            server_output_in_bytes = await asyncio.wait_for(server_process.stdout.readline(), 20)
+            server_output = server_output_in_bytes.decode()
+            full_server_output += server_output
+        except asyncio.TimeoutError:
+            logger.error(f"Timeout waiting for server startup message '{startup_message}'")
+            break
+        logger.debug(f"Server output: {server_output}")
+        if startup_message in server_output:
+            output_found = True
+            break
+
+    assert output_found, (
+        f"Full output:\n{full_server_output}\n"
+        + f"[ASSERT_ERROR] Startup log message '{startup_message}' not found in server output."
+    )
+
+    logger.info("Server started")
+
+    # Start n number of clients and capture their process objects
+    client_processes = []
+    for i in range(config["n_clients"]):
+        logger.info(f"Starting client {i}")
+        client_process = await asyncio.create_subprocess_exec(
+            "nohup",
+            "python",
+            "-m",
+            "examples.fedprox_example.client",
+            "--dataset_path",
+            dataset_path,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.STDOUT,
+        )
+        client_processes.append(client_process)
+
+    # Collecting the clients output when their processes finish
+    full_client_outputs = []
+    for i in range(len(client_processes)):
+        full_client_outputs.append(
+            await _wait_for_process_to_finish_and_retrieve_logs(client_processes[i], f"Client {i}")
+        )
+
+    logger.info("All clients finished execution")
+
+    # Collecting the server output when its process finish
+    full_server_output = await _wait_for_process_to_finish_and_retrieve_logs(server_process, "Server")
+
+    logger.info("Server has finished execution")
+
+    # server assertions
+    assert "error" not in full_server_output.lower(), (
+        f"Full output:\n{full_server_output}\n" + "[ASSERT ERROR] Error message found for server."
```

Comment:
Similar to the f-string above, I don't think you need to + part in these. They can either be part of the f-string proper or on a separate line and they'll work nicely 🙂

Reply:
If I break it in two different lines `black` will put it back into the same line because it's not over the line limit, and if I put it in a single string, `black` will format it like this:
```
assert (
    "error" not in full_server_output.lower()
), f"Full output:\n{full_server_output}\n[ASSERT ERROR] Error message found for server."
```
I honestly don't know what's worse. I have a slight preference for breaking the string in two. `black` is a bit too annoying on those edge cases.